File: lib/reflection.ts
import type { Reflection } from "@/types/journal";
import {
  buildPatternObservationsFromAnalysis,
  getPrimaryObservation,
} from "@/lib/observation-language";

/** Normalize reflections loaded from localStorage (legacy entries may omit new fields). */
export function normalizeReflection(
  raw: Partial<Reflection> & Pick<Reflection, "mood" | "emotionalIntensity">,
): Reflection {
  const themes = Array.isArray(raw.recurringThemes)
    ? raw.recurringThemes.filter((t): t is string => typeof t === "string")
    : [];

  const intensity = Number(raw.emotionalIntensity);

  const normalized: Reflection = {
    mood: String(raw.mood ?? "").trim(),
    emotionalIntensity: Math.min(
      10,
      Math.max(1, Number.isFinite(intensity) ? Math.round(intensity) : 5),
    ),
    recurringThemes: themes.slice(0, 4),
    hiddenConcern: String(raw.hiddenConcern ?? "").trim(),
    positiveSignal: String(raw.positiveSignal ?? "").trim(),
    recommendation: String(raw.recommendation ?? "").trim(),
    exactLanguagePattern: optionalString(raw.exactLanguagePattern),
    concreteObservation: optionalString(raw.concreteObservation),
    repeatedSignal: optionalString(raw.repeatedSignal),
    tensionOrContradiction: optionalString(raw.tensionOrContradiction),
    avoidedOrVagueArea: optionalString(raw.avoidedOrVagueArea),
    nextSmallAction: optionalString(raw.nextSmallAction),
    patternObservations: Array.isArray(raw.patternObservations)
      ? raw.patternObservations
          .filter((o): o is string => typeof o === "string")
          .map((o) => o.trim())
          .filter(Boolean)
          .slice(0, 6)
      : undefined,
    explainableConclusion: raw.explainableConclusion,
  };

  if (normalized.patternObservations && normalized.patternObservations.length > 0) {
    return normalized;
  }

  const derived = buildPatternObservationsFromAnalysis(normalized);
  if (derived.length > 0) {
    return { ...normalized, patternObservations: derived };
  }

  return normalized;
}

function optionalString(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

export interface SpecificReflectionView {
  exactLanguagePattern: string | null;
  concreteObservation: string;
  repeatedSignal: string | null;
  nextSmallAction: string;
  isLegacyFormat: boolean;
}

export function getSpecificReflectionView(
  reflection: Reflection,
): SpecificReflectionView {
  const hasNew =
    Boolean(reflection.exactLanguagePattern) ||
    Boolean(reflection.concreteObservation) ||
    Boolean(reflection.tensionOrContradiction) ||
    Boolean(reflection.repeatedSignal) ||
    Boolean(reflection.avoidedOrVagueArea) ||
    Boolean(reflection.nextSmallAction);

  return {
    exactLanguagePattern: reflection.exactLanguagePattern ?? null,
    concreteObservation:
      reflection.concreteObservation?.trim() ||
      "No concrete observation saved for this entry.",
    repeatedSignal:
      reflection.repeatedSignal?.trim() ||
      (reflection.recurringThemes.length > 0
        ? `Themes noted: ${reflection.recurringThemes.join(", ")}`
        : null),
    nextSmallAction: reflection.nextSmallAction?.trim() || "",
    isLegacyFormat: !hasNew,
  };
}

export function getEntryPreviewLine(reflection: Reflection): string {
  return getPrimaryObservation(reflection) ?? "Voice reflection";
}

File: apps/mobile/lib/models/reflection.dart
import 'package:archiveme_mobile/core/json/json_converters.dart';

class Reflection {
  const Reflection({
    required this.mood,
    required this.emotionalIntensity,
    required this.recurringThemes,
    required this.exactLanguagePattern,
    required this.concreteObservation,
    required this.repeatedSignal,
    this.tensionOrContradiction,
    this.avoidedOrVagueArea,
    this.nextSmallAction,
    this.patternObservations = const [],
  });

  factory Reflection.fromJson(Map<String, dynamic> json) {
    return Reflection(
      mood: JsonConverters.stringOrEmpty(json['mood']),
      emotionalIntensity: JsonConverters.intOrZero(json['emotionalIntensity']),
      recurringThemes: JsonConverters.stringList(json['recurringThemes']),
      exactLanguagePattern:
          JsonConverters.stringOrEmpty(json['exactLanguagePattern']),
      concreteObservation:
          JsonConverters.stringOrEmpty(json['concreteObservation']),
      repeatedSignal: JsonConverters.stringOrEmpty(json['repeatedSignal']),
      tensionOrContradiction:
          _optionalString(json['tensionOrContradiction']),
      avoidedOrVagueArea: _optionalString(json['avoidedOrVagueArea']),
      nextSmallAction: _optionalString(json['nextSmallAction']),
      patternObservations:
          JsonConverters.stringList(json['patternObservations']),
    );
  }

  final String mood;
  final int emotionalIntensity;
  final List<String> recurringThemes;
  final String exactLanguagePattern;
  final String concreteObservation;
  final String repeatedSignal;
  final String? tensionOrContradiction;
  final String? avoidedOrVagueArea;
  final String? nextSmallAction;
  final List<String> patternObservations;

  Map<String, dynamic> toJson() => {
        'mood': mood,
        'emotionalIntensity': emotionalIntensity,
        'recurringThemes': recurringThemes,
        'hiddenConcern': '',
        'positiveSignal': '',
        'recommendation': '',
        'exactLanguagePattern': exactLanguagePattern,
        'concreteObservation': concreteObservation,
        'repeatedSignal': repeatedSignal,
        if (tensionOrContradiction != null)
          'tensionOrContradiction': tensionOrContradiction,
        if (avoidedOrVagueArea != null) 'avoidedOrVagueArea': avoidedOrVagueArea,
        if (nextSmallAction != null) 'nextSmallAction': nextSmallAction,
        if (patternObservations.isNotEmpty)
          'patternObservations': patternObservations,
      };

  static String? _optionalString(Object? value) {
    if (value is! String) return null;
    final t = value.trim();
    return t.isEmpty ? null : t;
  }

  static bool _stringListsEqual(List<String> a, List<String> b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Reflection &&
          other.mood == mood &&
          other.emotionalIntensity == emotionalIntensity &&
          _stringListsEqual(other.recurringThemes, recurringThemes) &&
          other.exactLanguagePattern == exactLanguagePattern &&
          other.concreteObservation == concreteObservation &&
          other.repeatedSignal == repeatedSignal &&
          other.tensionOrContradiction == tensionOrContradiction &&
          other.avoidedOrVagueArea == avoidedOrVagueArea &&
          other.nextSmallAction == nextSmallAction &&
          _stringListsEqual(other.patternObservations, patternObservations);

  @override
  int get hashCode => Object.hash(
        mood,
        emotionalIntensity,
        Object.hashAll(recurringThemes),
        exactLanguagePattern,
        concreteObservation,
        repeatedSignal,
        tensionOrContradiction,
        avoidedOrVagueArea,
        nextSmallAction,
        Object.hashAll(patternObservations),
      );
}
